etl/command.py

```diff
@@ -270,22 +287,100 @@ def run_dag(
     if not force:
         print("--- Detecting which steps need rebuilding...")
         start_time = time.time()
-        steps = select_dirty_steps(steps, workers)
+        steps = select_dirty_steps(steps, workers=config.DIRTY_STEPS_WORKERS)
         click.echo(f"{click.style('OK', fg='blue')} ({time.time() - start_time:.1f}s)")
 
     if not steps:
         print("--- All datasets up to date!")
         return
 
-    print(f"--- Running {len(steps)} steps:")
+    if dry_run:
+        print(f"--- Running {len(steps)} steps:")
+        return enumerate_steps(steps)
+    elif workers == 1:
+        print(f"--- Running {len(steps)} steps:")
+        return exec_steps(steps, strict=strict)
+    else:
+        print(f"--- Running {len(steps)} steps with {workers} processes:")
+        return exec_steps_parallel(steps, workers, dag=dag, strict=strict)
+
+
+def exec_steps(steps: List[Step], strict: Optional[bool] = None) -> None:
     for i, step in enumerate(steps, 1):
         print(f"--- {i}. {step}...")
-        if not dry_run:
-            strict = _detect_strictness_level(step, strict)
-            with strictness_level(strict):
-                time_taken = timed_run(lambda: step.run())
-                click.echo(f"{click.style('OK', fg='blue')} ({time_taken:.1f}s)")
-                print()
+        strict = _detect_strictness_level(step, strict)
+        with strictness_level(strict):
+            time_taken = timed_run(lambda: step.run())
+            click.echo(f"{click.style('OK', fg='blue')} ({time_taken:.1f}s)")
+            print()
+
+
+def exec_steps_parallel(steps: List[Step], workers: int, dag: DAG, strict: Optional[bool] = None) -> None:
+    # put grapher steps in front of the queue to process them as soon as possible and lessen
+    # the load on MySQL
+    steps = sorted(steps, key=lambda x: "grapher://" not in str(x))
+
+    # create execution graph from steps
+    exec_graph = {}
+    steps_str = {str(step) for step in steps}
+    for step in steps:
+        # only add dependencies that are in the list of steps (i.e. are dirty)
+        # NOTE: we have to compare their string versions, the actual objects might have
+        # different attributes
+        exec_graph[str(step)] = {str(dep) for dep in step.dependencies if str(dep) in steps_str}
+
+    exec_graph_parallel(exec_graph, _exec_step_job, workers, dag=dag, strict=strict)
+
+
+def exec_graph_parallel(
+    exec_graph: Dict[str, Any], func: Callable[[str], None], workers: int, use_threads=False, **kwargs
+) -> None:
+    # Create a topological sorter and add tasks with their dependencies
```

Comment:
Could you please briefly explain the logic of this sorter?

Reply:
Good point, added.